taraxa/state/dpos_2.0/precompiled/delegations.go

```diff
@@ -0,0 +1,123 @@
+	if delegation == nil {
+		panic("ModifyDelegation: delegation cannot be nil")
+	}
+
+	delegator_validators := self.getDelegatorValidatorsList(delegator_address)
+
+	// This could happen only due to some serious logic bug
+	if delegator_validators.AccountExists(validator_address) == false {
+		panic("ModifyDelegation: non existent delegation")
+	}
+
+	key := self.genDelegationKey(delegator_address, validator_address)
+	self.storage.Put(key, rlp.MustEncodeToBytes(delegation))
+}
+
+func (self *Delegations) CreateDelegation(delegator_address *common.Address, validator_address *common.Address, block types.BlockNum, stake *big.Int) {
+	// Creates Delegation object in storage
+	delegation := new(Delegation)
+	delegation.Stake = stake
+	delegation.LastUpdated = block
+
+	delegation_key := self.genDelegationKey(delegator_address, validator_address)
+	self.storage.Put(delegation_key, rlp.MustEncodeToBytes(delegation))
+
+	// Adds validator into delegator's validators list
+	delegator_validators := self.getDelegatorValidatorsList(delegator_address)
+	delegator_validators.CreateAccount(validator_address)
+}
+
+func (self *Delegations) RemoveDelegation(delegator_address *common.Address, validator_address *common.Address) {
+	delegation_key := self.genDelegationKey(delegator_address, validator_address)
+	self.storage.Put(delegation_key, nil)
+
+	// Removes validator from delegator's validators list
+	delegator_validators := self.getDelegatorValidatorsList(delegator_address)
+	remaining_validators_count := delegator_validators.RemoveAccount(validator_address)
+
+	// If delegator does not delegate to anyone else, remove also his validators list
+	if remaining_validators_count == 0 {
+		self.removeDelegatorValidatorsList(delegator_address)
+	}
+}
+
+func (self *Delegations) GetDelegatorValidatorsAddresses(delegator_address *common.Address, batch uint32, count uint32) ([]common.Address, bool) {
+	delegator_validators := self.getDelegatorValidatorsList(delegator_address)
+	return delegator_validators.GetAccounts(batch, count)
+}
+
+func (self *Delegations) getDelegatorValidatorsList(delegator_address *common.Address) *IterableMap {
+	delegator_validators, found := self.delegators_validators[*delegator_address]
+	if found == false {
+		delegator_validators = new(IterableMap)
+		delegator_validators_field := append(self.delegators_validators_field_prefix, delegator_address[:]...)
+		delegator_validators.Init(self.storage, delegator_validators_field)
+	}
+
+	return delegator_validators
+}
+
+func (self *Delegations) removeDelegatorValidatorsList(delegator_address *common.Address) {
+	delete(self.delegators_validators, *delegator_address)
+}
+
+func (self *Delegations) genDelegationKey(delegator_address *common.Address, validator_address *common.Address) *common.Hash {
+	// TODO: IMPORTANT -> is there an reason why to use stor_k_2 vs stor_k_1 ???
```

Comment:
stor_k_2 returns by value, so I believe it should be also used here as the value

Reply:
as we are not returning key it is ok to use this version